src/tree/mod.rs

```diff
@@ -1,103 +1,343 @@
 pub mod block_scanner;
 pub mod config;
 pub mod error;
+pub mod identity_tree;
 pub mod service;
-pub mod tree_data;
-pub mod tree_updater;
+pub mod tree_manager;
 
+use crate::error::Log as _;
+use std::collections::{HashMap, HashSet};
 use std::sync::atomic::{AtomicBool, Ordering};
 use std::sync::Arc;
-use std::time::Duration;
 
-use error::TreeAvailabilityError;
 use ethers::providers::Middleware;
-use ethers::types::H160;
-use semaphore::lazy_merkle_tree::{Canonical, LazyMerkleTree};
+use ethers::types::U256;
+use ruint::Uint;
+use semaphore::dynamic_merkle_tree::DynamicMerkleTree;
+use semaphore::lazy_merkle_tree::LazyMerkleTree;
 use semaphore::merkle_tree::Hasher;
 use semaphore::poseidon_tree::PoseidonHash;
+use tokio::sync::mpsc::Receiver;
 use tokio::sync::RwLock;
 use tokio::task::JoinHandle;
+use tokio::time::Instant;
 use tracing::instrument;
 
-use self::tree_data::TreeData;
-use self::tree_updater::TreeUpdater;
+use self::identity_tree::{IdentityTree, InclusionProof, LeafUpdates, Root};
+use self::tree_manager::{
+    extract_identity_updates, BridgedTree, CanonicalTree, TreeManager,
+};
+use crate::abi::IBridgedWorldID;
+use crate::tree::identity_tree::flatten_leaf_updates;
 
 pub type PoseidonTree<Version> = LazyMerkleTree<PoseidonHash, Version>;
 pub type Hash = <PoseidonHash as Hasher>::Hash;
 
-pub const SYNC_TO_HEAD_SLEEP_SECONDS: u64 = 5;
-
-/// An abstraction over a tree with a history of changes
-///
-/// In our data model the `tree` is the oldest available tree.
-/// The entires in `tree_history` represent new additions to the tree.
 pub struct WorldTree<M: Middleware> {
```

Comment:
Would be awesome to get some comments here describing how all of these pieces connect together